paper-api/src/main/java/io/papermc/paper/event/player/AbstractRespawnEvent.java

```diff
@@ -0,0 +1,82 @@
+package io.papermc.paper.event.player;
+
+import com.google.common.collect.ImmutableSet;
+import org.bukkit.Location;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerEvent;
+import org.bukkit.event.player.PlayerRespawnEvent;
+import org.jspecify.annotations.NullMarked;
+import java.util.Set;
+
+@NullMarked
+public abstract class AbstractRespawnEvent extends PlayerEvent {
+
+    protected Location respawnLocation;
+    private final boolean isBedSpawn;
+    private final boolean isAnchorSpawn;
+    private final PlayerRespawnEvent.RespawnReason respawnReason;
+    private final Set<PlayerRespawnEvent.RespawnFlag> respawnFlags;
+
+    protected AbstractRespawnEvent(
+        final Player respawnPlayer, final Location respawnLocation, final boolean isBedSpawn,
+        final boolean isAnchorSpawn, final PlayerRespawnEvent.RespawnReason respawnReason
+    ) {
+        super(respawnPlayer);
+        this.respawnLocation = respawnLocation;
+        this.isBedSpawn = isBedSpawn;
+        this.isAnchorSpawn = isAnchorSpawn;
+        this.respawnReason = respawnReason;
+        ImmutableSet.Builder<PlayerRespawnEvent.RespawnFlag> builder = ImmutableSet.builder();
+        if (respawnReason == PlayerRespawnEvent.RespawnReason.END_PORTAL) builder.add(PlayerRespawnEvent.RespawnFlag.END_PORTAL);
+        if (this.isBedSpawn) builder.add(PlayerRespawnEvent.RespawnFlag.BED_SPAWN);
+        if (this.isAnchorSpawn) builder.add(PlayerRespawnEvent.RespawnFlag.ANCHOR_SPAWN);
+        this.respawnFlags = builder.build();
+    }
+
+    /**
+     * Gets the current respawn location
+     *
+     * @return the current respawn location
+     */
+    public Location getRespawnLocation() {
+        return this.respawnLocation;
```

Comment:
This should be cloned if the server isn't using this value after the event is fired.